### PR TITLE
test: E2E: goroutine and err response cleanup

### PR DESCRIPTION
**Pull request by jackfrancis**
**Reason for Change**:

This PR began as:

"We shouldn't be intolerant towards ephemeral errors in "wait for state" tests."

And it evolved into a wholesale cleanup of background/async/retry go func usage patterns:

- No longer holding open unnecessary channels for goroutine state communication (e.g., we were commonly holding open a distinct error channel, and a distinct "response" channel.
- Converged lots of "do this then do this if not err then do this if not err" into simpler "do stuff and err immediately" wrapped inside simpler "tolerate errs and retry until timeout" funcs.
- Cleaned up some var naming so that it's clear we're doing "retry" and then "time out" implementations (e.g., rename the ill-worded `duration` var to `timeout`
- Various, targeted retry/timeout changes in response to deep investigation of tests that rely upon error tolerance + retry behaviors.
- Lots more "describe resource" after errors occur

Extra credit:

- Ensured that we configure "fast" HPA config for all E2E-consuming api model, including those with no opinionated k8s version

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by jackfrancis** [comment]
Addresses this flake here:

https://dev.azure.com/aks-engine/AKS-Engine/_build/results?buildId=2387&view=logs&jobId=d00b8625-2481-5061-ccc1-3c7b5b6311ba

@mboersma @devigned FYI

**Comment by jackfrancis** [comment]
And here: https://dev.azure.com/aks-engine/AKS-Engine/_build/results?buildId=2387&view=logs

**Comment by jackfrancis** [comment]
Sorry @devigned, I found more instances of this pattern :/

**Comment by devigned** [comment]
Do you want to keep the error message or output the error message from these as they loop? I'm asking out of concern for determining the root issue from the build log.

**Comment by jackfrancis** [comment]
I think I just want us to bubble up the originating error, not add our own curated stdout.

**Comment by jackfrancis** [comment]
... And we're not doing that.

Let's not merge this super quickly (unless we see a bunch more flakes). It's definitely wrong to return immediately for these particular errors, but clearly we're not managing them well.

File: test/e2e/kubernetes/job/job.go [hunk 1]
```diff
@@ -206,19 +206,29 @@ func AreAllJobsCompleted(jobPrefix, namespace string) (bool, error) {
 	return true, nil
 }
 
-// WaitOnReady is used when you dont have a handle on a job but want to wait until its in a Succeeded state.
-func WaitOnReady(jobPrefix, namespace string, sleep, duration time.Duration) (bool, error) {
+// WaitOnSucceeded returns true if all jobs matching a prefix substring are in a succeeded state within a period of time
+func WaitOnSucceeded(jobPrefix, namespace string, sleep, duration time.Duration) (bool, error) {
 	readyCh := make(chan bool, 1)
 	errCh := make(chan error)
```
Comment (devigned) [inline test/e2e/kubernetes/job/job.go]: Could probably take these and turn them into one return struct with one channel.

File: test/e2e/kubernetes/deployment/deployment.go [hunk 2]
```diff
@@ -387,47 +387,42 @@ func (d *Deployment) Pods() ([]pod.Pod, error) {
...
 		}
 	}()
 	for {
 		select {
-		case err := <-errCh:
-			return pods, err
-		case <-readyCh:
-			return pods, nil
+		case result := <-ch:
+			mostRecentWaitForReplicasError = result.Err
```
[the remark this answers is not in the document]
Comment (jackfrancis) [reply test/e2e/kubernetes/deployment/deployment.go]: I assume you mean we might wrap lines 411-423 around a `if mostRecentWaitForReplicasError == nil` block?

File: test/e2e/kubernetes/job/job.go [hunk 3]
```diff
@@ -206,19 +206,29 @@ func AreAllJobsCompleted(jobPrefix, namespace string) (bool, error) {
 	return true, nil
 }
 
-// WaitOnReady is used when you dont have a handle on a job but want to wait until its in a Succeeded state.
-func WaitOnReady(jobPrefix, namespace string, sleep, duration time.Duration) (bool, error) {
+// WaitOnSucceeded returns true if all jobs matching a prefix substring are in a succeeded state within a period of time
+func WaitOnSucceeded(jobPrefix, namespace string, sleep, duration time.Duration) (bool, error) {
```
[the remark this answers is not in the document]
Comment (jackfrancis) [reply test/e2e/kubernetes/job/job.go]: yes, I'm trying to standardize all of these to `timeout` (which I think is clearer, will fix).

File: test/e2e/kubernetes/deployment/deployment.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (devigned) [reply test/e2e/kubernetes/deployment/deployment.go]: Pretty much. I was going to suggest a short circuit for the select in the case of an error.
```go
if mostRecentWaitForReplicasError != nil {
  break
}

File: test/e2e/kubernetes/kubernetes_test.go [hunk 4]
```diff
@@ -1415,7 +1440,13 @@ var _ = Describe("Azure Container Cluster using the Kubernetes Orchestrator", fu
 
 				By("Ensuring we no longer have ingress access from the network-policy pods to backend pods")
 				for _, backendPod := range backendPods {
-					pass, err = pl.ValidateCurlConnection(backendPod.Status.PodIP, 5*time.Second, validateNetworkPolicyTimeout)
+					pass, err = pl.ValidateCurlConnection(backendPod.Status.PodIP, 30*time.Second, validateNetworkPolicyTimeout)
+					if err != nil {
+						e := backendPod.Describe()
+						if e != nil {
```
Comment (devigned) [inline test/e2e/kubernetes/kubernetes_test.go]: Great to see more output on errors!

**Pull request merged**